File: hello-websocket-csharp/test/CodecTest.cs
using HelloWebSocket;
using Xunit;

namespace CodecTest;

public class CodecTest
{
    [Fact]
    public void TestHelloWorkedExample()
    {
        var msg = Codec.Hello("Go");
        var data = msg.Encode();
        byte[] expected = [0x48, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x06, 0x00, 0x00, 0x00, 0x02, 0x47, 0x6F];
        Assert.Equal(expected, data);
    }

    [Fact]
    public void TestRoundTripAll()
    {
        var messages = new Codec.Message[]
        {
            Codec.Hello("C#"),
            Codec.Bonjour("Go"),
            Codec.EchoRequestMsg(42, "Python", "hello"),
            Codec.Ping(1700000000000),
            Codec.Pong(1700000000001),
            Codec.TimeNotif(1700000000000, "2023-11-14T22:13:20Z"),
            Codec.RandomNumberMsg(99, 42),
            Codec.HashResponseMsg(99, "7688b6ef5a"),
            Codec.DisconnectMsg("bye"),
            Codec.ErrorMsg(Codec.ERR_UNKNOWN_MSG_TYPE, "bad type"),
        };
        foreach (var orig in messages)
        {
            var decoded = Codec.DecodeMessage(orig.Encode());
            Assert.Equal(orig.Type, decoded.Type);
        }
    }

    [Fact]
    public void TestRoundTripEchoResponse()
    {
        var kv = new Dictionary<string, string> { ["id"] = "1", ["data"] = "Hello" };
        var orig = Codec.EchoResponseMsg(200, new[] { new Codec.EchoResult(123, 0, kv) });
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_ECHO_RESPONSE, decoded.Type);
        Assert.Equal(200, decoded.EchoStatus);
        Assert.Single(decoded.EchoResults!);
        Assert.Equal("1", decoded.EchoResults![0].Kv["id"]);
    }

    [Fact]
    public void TestRoundTripKiss()
    {
        var orig = Codec.KissRequest("Linux", "6.6", "arch", "AMD64");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_KISS_REQUEST, decoded.Type);
        Assert.Equal("Linux", decoded.OsName);
        Assert.Equal("AMD64", decoded.OsArch);
    }

    [Fact]
    public void TestBadMagic()
    {
        var data = new byte[] { 0x00, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00 };
        Assert.Throws<Exception>(() => Codec.DecodeFrame(data));
    }

    [Fact]
    public void TestBadVersion()
    {
        var data = new byte[] { 0x48, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00 };
        Assert.Throws<Exception>(() => Codec.DecodeFrame(data));
    }

    [Fact]
    public void TestTruncatedPayload()
    {
        var data = new byte[] { 0x48, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0xFF };
        Assert.Throws<Exception>(() => Codec.DecodeFrame(data));
    }

    [Fact]
    public void TestHashNumber()
    {
        var h = Codec.HashNumber(42);
        Assert.Equal(10, h.Length);
        Assert.Equal(h, Codec.HashNumber(42));
    }

    [Fact]
    public void TestRoundTripKissResponse()
    {
        var orig = Codec.KissResponse("en_US", "UTF-8", "UTC");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_KISS_RESPONSE, decoded.Type);
        Assert.Equal("en_US", decoded.KissLanguage);
        Assert.Equal("UTF-8", decoded.KissEncoding);
        Assert.Equal("UTC", decoded.KissTimeZone);
    }

    [Fact]
    public void TestRoundTripDisconnect()
    {
        var orig = Codec.DisconnectMsg("test reason");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_DISCONNECT, decoded.Type);
        Assert.Equal("test reason", decoded.DisconnectReason);
    }

    [Fact]
    public void TestRoundTripError()
    {
        var orig = Codec.ErrorMsg(Codec.ERR_DECODE, "decode failed");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_ERROR, decoded.Type);
        Assert.Equal(Codec.ERR_DECODE, decoded.ErrorCode);
        Assert.Equal("decode failed", decoded.ErrorMessage);
    }

    [Fact]
    public void TestRoundTripRandomNumber()
    {
        var orig = Codec.RandomNumberMsg(5, 99999);
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_RANDOM_NUMBER, decoded.Type);
        Assert.Equal(5, decoded.RandomId);
        Assert.Equal(99999, decoded.RandomNumber);
    }

    [Fact]
    public void TestRoundTripHashResponse()
    {
        var orig = Codec.HashResponseMsg(7, "abcdef1234");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(Codec.MSG_HASH_RESPONSE, decoded.Type);
        Assert.Equal(7, decoded.RandomId);
        Assert.Equal("abcdef1234", decoded.HashHex);
    }

    [Fact]
    public void TestEmptyString()
    {
        var orig = Codec.DisconnectMsg("");
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal("", decoded.DisconnectReason);
    }

    [Fact]
    public void TestUnknownMsgType()
    {
        var data = new byte[] { 0x48, 0x01, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00 };
        Assert.Throws<Exception>(() => Codec.DecodeMessage(data));
    }

    [Fact]
    public void TestEmptyEchoResults()
    {
        var orig = Codec.EchoResponseMsg(204, Array.Empty<Codec.EchoResult>());
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(204, decoded.EchoStatus);
        Assert.Empty(decoded.EchoResults!);
    }

    [Fact]
    public void TestMultipleEchoResults()
    {
        var results = new[]
        {
            new Codec.EchoResult(1, 0, new Dictionary<string, string> { ["a"] = "1" }),
            new Codec.EchoResult(2, 1, new Dictionary<string, string> { ["b"] = "2" }),
        };
        var orig = Codec.EchoResponseMsg(200, results);
        var decoded = Codec.DecodeMessage(orig.Encode());
        Assert.Equal(2, decoded.EchoResults!.Length);
        Assert.Equal("1", decoded.EchoResults[0].Kv["a"]);
        Assert.Equal("2", decoded.EchoResults[1].Kv["b"]);
    }
}

File: hello-websocket-java/common/src/test/java/io/github/hellowebsocket/CodecTest.java
package io.github.hellowebsocket;

import static org.junit.jupiter.api.Assertions.*;

import java.util.HashMap;
import java.util.Map;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;

class CodecTest {

    @Test
    @DisplayName("HELLO worked example byte-level match")
    void testHelloWorkedExample() {
        Codec.Message msg = Codec.hello("Go");
        byte[] data = msg.encode();
        byte[] expected = {
            0x48, 0x01, 0x01, 0x00,
            0x00, 0x00, 0x00, 0x06,
            0x00, 0x00, 0x00, 0x02,
            0x47, 0x6F
        };
        assertArrayEquals(expected, data);
    }

    @Test
    @DisplayName("Round-trip all simple message types")
    void testRoundTripAll() throws Codec.DecodeException {
        Codec.Message[] messages = {
            Codec.hello("Java"),
            Codec.bonjour("Go"),
            Codec.echoRequest(42, "Python", "hello"),
            Codec.ping(1700000000000L),
            Codec.pong(1700000000001L),
            Codec.timeNotif(1700000000000L, "2023-11-14T22:13:20Z"),
            Codec.randomNumber(99, 42),
            Codec.hashResponse(99, "7688b6ef5a"),
            Codec.disconnect("bye"),
            Codec.error(Codec.ERR_UNKNOWN_MSG_TYPE, "bad type"),
        };
        for (Codec.Message orig : messages) {
            Codec.Message decoded = Codec.decodeMessage(orig.encode());
            assertEquals(orig.type, decoded.type);
        }
    }

    @Test
    @DisplayName("Round-trip EchoResponse")
    void testRoundTripEchoResponse() throws Codec.DecodeException {
        Map<String, String> kv = new HashMap<>();
        kv.put("id", "1");
        kv.put("data", "Hello");
        Codec.EchoResult result = new Codec.EchoResult(123, 0, kv);
        Codec.Message orig = Codec.echoResponse(200, new Codec.EchoResult[]{result});
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_ECHO_RESPONSE, decoded.type);
        assertEquals(200, decoded.echoStatus);
        assertEquals(1, decoded.echoResults.length);
        assertEquals("1", decoded.echoResults[0].kv().get("id"));
    }

    @Test
    @DisplayName("Round-trip Kiss")
    void testRoundTripKiss() throws Codec.DecodeException {
        Codec.Message orig = Codec.kissRequest("Linux", "6.6", "arch", "AMD64");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_KISS_REQUEST, decoded.type);
        assertEquals("Linux", decoded.osName);
        assertEquals("AMD64", decoded.osArch);
    }

    @Test
    @DisplayName("Bad magic rejected")
    void testBadMagic() {
        byte[] data = {0x00, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00};
        assertThrows(Codec.DecodeException.class, () -> Codec.decodeFrame(data));
    }

    @Test
    @DisplayName("Bad version rejected")
    void testBadVersion() {
        byte[] data = {0x48, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00};
        assertThrows(Codec.DecodeException.class, () -> Codec.decodeFrame(data));
    }

    @Test
    @DisplayName("Truncated payload rejected")
    void testTruncatedPayload() {
        byte[] data = {0x48, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, (byte)0xFF};
        assertThrows(Codec.DecodeException.class, () -> Codec.decodeFrame(data));
    }

    @Test
    @DisplayName("Hash number produces 10-char hex")
    void testHashNumber() {
        String h = Codec.hashNumber(42);
        assertEquals(10, h.length());
        assertEquals(h, Codec.hashNumber(42));
    }

    @Test
    @DisplayName("Round-trip KissResponse")
    void testRoundTripKissResponse() throws Codec.DecodeException {
        Codec.Message orig = Codec.kissResponse("en_US", "UTF-8", "UTC");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_KISS_RESPONSE, decoded.type);
        assertEquals("en_US", decoded.kissLanguage);
        assertEquals("UTF-8", decoded.kissEncoding);
        assertEquals("UTC", decoded.kissTimeZone);
    }

    @Test
    @DisplayName("Round-trip Disconnect")
    void testRoundTripDisconnect() throws Codec.DecodeException {
        Codec.Message orig = Codec.disconnect("test reason");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_DISCONNECT, decoded.type);
        assertEquals("test reason", decoded.disconnectReason);
    }

    @Test
    @DisplayName("Round-trip Error")
    void testRoundTripError() throws Codec.DecodeException {
        Codec.Message orig = Codec.error(Codec.ERR_DECODE, "decode failed");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_ERROR, decoded.type);
        assertEquals(Codec.ERR_DECODE, decoded.errorCode);
        assertEquals("decode failed", decoded.errorMessage);
    }

    @Test
    @DisplayName("Round-trip RandomNumber")
    void testRoundTripRandomNumber() throws Codec.DecodeException {
        Codec.Message orig = Codec.randomNumber(5, 99999);
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_RANDOM_NUMBER, decoded.type);
        assertEquals(5, decoded.randomId);
        assertEquals(99999, decoded.randomNumber);
    }

    @Test
    @DisplayName("Round-trip HashResponse")
    void testRoundTripHashResponse() throws Codec.DecodeException {
        Codec.Message orig = Codec.hashResponse(7, "abcdef1234");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(Codec.MSG_HASH_RESPONSE, decoded.type);
        assertEquals(7, decoded.randomId);
        assertEquals("abcdef1234", decoded.hashHex);
    }

    @Test
    @DisplayName("Empty string round-trip")
    void testEmptyString() throws Codec.DecodeException {
        Codec.Message orig = Codec.disconnect("");
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals("", decoded.disconnectReason);
    }

    @Test
    @DisplayName("Unknown message type rejected")
    void testUnknownMsgType() {
        byte[] data = {0x48, 0x01, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00};
        assertThrows(Codec.DecodeException.class, () -> Codec.decodeMessage(data));
    }

    @Test
    @DisplayName("Empty EchoResponse results array")
    void testEmptyEchoResults() throws Codec.DecodeException {
        Codec.Message orig = Codec.echoResponse(204, new Codec.EchoResult[]{});
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(204, decoded.echoStatus);
        assertEquals(0, decoded.echoResults.length);
    }

    @Test
    @DisplayName("EchoResponse with multiple results")
    void testMultipleEchoResults() throws Codec.DecodeException {
        Map<String, String> kv1 = new HashMap<>();
        kv1.put("a", "1");
        Map<String, String> kv2 = new HashMap<>();
        kv2.put("b", "2");
        Codec.EchoResult[] results = {
            new Codec.EchoResult(1, 0, kv1),
            new Codec.EchoResult(2, 1, kv2),
        };
        Codec.Message orig = Codec.echoResponse(200, results);
        Codec.Message decoded = Codec.decodeMessage(orig.encode());
        assertEquals(2, decoded.echoResults.length);
        assertEquals("1", decoded.echoResults[0].kv().get("a"));
        assertEquals("2", decoded.echoResults[1].kv().get("b"));
    }
}
